rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/metadata/CalculateDeltaNode.java

```diff
@@ -0,0 +1,175 @@
+                        }
+
+                        ObjectNode result = (ObjectNode) json;
+                        result.put(config.getOutputValueKey(), delta);
+
+                        if (config.isAddPeriodBetweenMsgs()) {
+                            long period = previousData != null ? currentTs - previousData.ts : 0;
+                            result.put(config.getPeriodValueKey(), period);
+                        }
+                        ctx.tellSuccess(TbMsg.transformMsg(msg, msg.getType(), msg.getOriginator(), msg.getMetaData(), JacksonUtil.toString(result)));
+                    },
+                    t -> ctx.tellFailure(msg, t)
+                    , ctx.getDbCallbackExecutor());
+        } else if (config.isTellFailureIfInputValueKeyIsAbsent()) {
+            ctx.tellNext(msg, TbRelationTypes.FAILURE);
+        } else {
+            ctx.tellSuccess(msg);
+        }
+    }
+
+    @Override
+    public void destroy() {
+
+    }
+
+    private ListenableFuture<ValueWithTs> fetchLatestValue(EntityId entityId) {
+        return Futures.transform(timeseriesService.findLatest(ctx.getTenantId(), entityId, Collections.singletonList(config.getInputValueKey())),
+                list -> extractValue(list.get(0))
+                , ctx.getDbCallbackExecutor());
+    }
+
+    private ListenableFuture<ValueWithTs> getLastValue(EntityId entityId) {
+        ValueWithTs latestValue;
+        if (useCache && (latestValue = cache.get(entityId)) != null) {
+            return Futures.immediateFuture(latestValue);
+        } else {
+            return fetchLatestValue(entityId);
+        }
+    }
+
+    private ValueWithTs extractValue(TsKvEntry kvEntry) {
+        double result = 0.0;
+        long ts = 0;
+        if (kvEntry != null) {
+            ts = kvEntry.getTs();
+            switch (kvEntry.getDataType()) {
+                case LONG:
+                    result = kvEntry.getLongValue().get();
+                    break;
+                case DOUBLE:
+                    result = kvEntry.getDoubleValue().get();
+                    break;
+                case BOOLEAN:
+                    result = kvEntry.getBooleanValue().get() ? 1 : 0;
+                    break;
+                case STRING:
+                    String str = kvEntry.getStrValue().orElse(null);
+                    try {
+                        if (str == null) {
+                            return null;
+                        }
+                        result = Double.parseDouble(str);
+                    } catch (NumberFormatException e) {
+                        throw new IllegalArgumentException("Calculation failed. Unable to parse value [" + str + "]" +
+                                " of telemetry [" + kvEntry.getKey() + "] to Double");
+                    }
+                    break;
+            }
+        }
+        return new ValueWithTs(ts, result);
```

Comment:
So, if there was no previous reading, we will calculate the difference between the current reading and 0. This is wrong.